src/lightning/fabric/plugins/precision/bitsandbytes.py

```diff
@@ -164,11 +168,36 @@ def _quantize_on_load_hook(quantize_fn: Callable[[torch.Tensor], None], state_di
 
 
 def _ignore_missing_weights_hook(module: torch.nn.Module, incompatible_keys: _IncompatibleKeys) -> None:
+    # since we manually loaded the weight in the `_quantize_on_load_hook` hook, we need to avoid this missing key false
+    # positive
     for key in reversed(incompatible_keys.missing_keys):
         if key.endswith("weight"):
             incompatible_keys.missing_keys.remove(key)
 
 
+def _replace_param(
+    p: torch.nn.Parameter, data: torch.Tensor, quant_state: Optional[Tuple] = None
+) -> torch.nn.Parameter:
+    bnb = _import_bitsandbytes()
+
+    # doing `param.data = weight` raises a RuntimeError if param.data was on meta-device, so
+    # we need to re-create the parameters instead of overwriting the data
+    if p.device.type == "meta":
+        if isinstance(p, bnb.nn.Params4bit):
+            return bnb.nn.Params4bit(
+                data,
+                requires_grad=data.requires_grad,
+                quant_state=quant_state,
+                compress_statistics=p.compress_statistics,
+                quant_type=p.quant_type,
+            )
+        return torch.nn.Parameter(data, requires_grad=data.requires_grad)
+    p.data = data
+    if isinstance(p, bnb.nn.Params4bit):
+        p.quant_state = quant_state
+    return p
```

Comment:
```suggestion
    param: torch.nn.Parameter, data: torch.Tensor, quant_state: Optional[Tuple] = None
) -> torch.nn.Parameter:
    bnb = _import_bitsandbytes()

    # doing `param.data = weight` raises a RuntimeError if param.data was on meta-device, so
    # we need to re-create the parameters instead of overwriting the data
    if param.device.type == "meta":
        if isinstance(param, bnb.nn.Params4bit):
            return bnb.nn.Params4bit(
                data,
                requires_grad=data.requires_grad,
                quant_state=quant_state,
                compress_statistics=p.compress_statistics,
                quant_type=p.quant_type,
            )
        return torch.nn.Parameter(data, requires_grad=data.requires_grad)
    param.data = data
    if isinstance(param, bnb.nn.Params4bit):
        param.quant_state = quant_state
    return param
```